caikit_nlp/modules/text_embedding/crossencoder.py

```diff
@@ -0,0 +1,807 @@
+        )
+        artifacts_path = self._ARTIFACTS_PATH_DEFAULT
+        saver.update_config({self._ARTIFACTS_PATH_KEY: artifacts_path})
+
+        # Save the model
+        self.model.save(os.path.join(model_config_path, artifacts_path))
+
+        # Save the config
+        ModuleConfig(saver.config).save(model_config_path)
+
+
+class CrossEncoderWithTruncate(CrossEncoder):
+    def __init__(
+        self,
+        model_name: str,
+        num_labels: int = None,
+        max_length: int = None,
+        device: str = None,
+        tokenizer_args: Dict = None,
+        automodel_args: Dict = None,
+        trust_remote_code: bool = False,
+        revision: Optional[str] = None,
+        local_files_only: bool = False,
+        default_activation_function=None,
+        classifier_dropout: float = None,
+    ):
+        super().__init__(
+            model_name,
+            num_labels,
+            max_length,
+            device,
+            tokenizer_args,
+            automodel_args,
+            trust_remote_code,
+            revision,
+            local_files_only,
+            default_activation_function,
+            classifier_dropout,
+        )
+        self.tokenizers = {}
+
+    def _get_tokenizer_per_thread(self):
+        """Use a copy of the tokenizer per-model (self) and per-thread (map by thread ID)."""
+
+        # Keep copies of tokenizer per thread (in each wrapped model instance)
+        thread_id = threading.get_ident()
+        tokenizer = (
+            self.tokenizers[thread_id]
+            if thread_id in self.tokenizers
+            else self.tokenizers.setdefault(thread_id, deepcopy(self.tokenizer))
+        )
+
+        return tokenizer
+
+    def get_tokenized(self, texts, **kwargs):
+        """Intentionally always call tokenizer the same way to avoid thread issues.
+
+        Use a copy of the tokenizer per-model (self) and per-thread (map by thread ID).
+
+        Avoid changing the max length, truncation, and padding to avoid the
+        "Already borrowed" errors that come with concurrent threads attempting to use
+        the fast tokenizer with different truncation settings.
+        """
+
+        max_len = kwargs.get("truncate_input_tokens", self.tokenizer.model_max_length)
+        max_len = min(max_len, self.tokenizer.model_max_length)
+        if max_len <= 0:
+            max_len = None
+
+        tokenizer = self._get_tokenizer_per_thread()
+        tokenized = tokenizer(
+            *texts,
+            return_attention_mask=True,  # Used for determining token count
+            return_token_type_ids=False,  # Needed for cross-encoders
+            return_overflowing_tokens=False,  # DO NOT USE overflow tokens break sentence batches
+            return_offsets_mapping=True,  # Used for truncation test
+            return_length=False,
+            return_tensors="pt",
+            truncation=True,
+            padding=True,
+            max_length=max_len,
```

Comment:
Good catch, but after all that "do not touch this" strategy, Nick suggested we should have separate tokenizers per thread anyway.  Following that advice here (because he's probably right) and now we no longer get already borrowed issues when changing a param on a thread-dedicated tokenizer and don't need to write another truncation strategy (pair truncation).

I tested this with a multi-threaded client using randomized truncation to make sure I see the error before going tokenizer-per-thread and don't see it after.  I don't think that test fits our unit tests(?).

Reply:
oops, I see the docstring comments now.   Fixed